src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs

```diff
@@ -418,87 +373,50 @@ private static bool TryValidateHost(string host)
             return false;
         }
 
-        var i = 0;
-        for (; i < host.Length; i++)
+        var firstNonHostCharIdx = host.AsSpan().IndexOfAnyExcept(HostChars);
+        if (firstNonHostCharIdx == -1)
         {
-            if (!IsValidHostChar(host[i]))
-            {
-                break;
-            }
+            // no port
+            return true;
+        }
+        else
+        {
+            return TryValidateHostPort(host, firstNonHostCharIdx);
         }
-        return TryValidateHostPort(host, i);
-    }
-
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    private static bool IsValidHostChar(char ch)
-    {
-        return ch < HostCharValidity.Length && HostCharValidity[ch];
     }
 
     // The lead '[' was already checked
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
     private static bool TryValidateIPv6Host(string hostText)
     {
-        for (var i = 1; i < hostText.Length; i++)
+        var hostEndIdx = hostText.IndexOf(']');
+        if (hostEndIdx == -1 || hostEndIdx < 4)
```

Comment:
May be slightly cheaper for the common case with something like
```c#
ReadOnlySpan<char> host = hostText.Slice(1);
int hostEndIdx = host.IndexOfAnyExcept(Ipv6HostChars); // Doubles as an IndexOf(']')
if ((uint)hostEndIdx >= (uint)host.Length || host[hostEndIdx] != ']')
{
    return false;
}

return (hostEndIdx + 1 == hostText.Length) || TryValidateHostPort(hostText, hostEndIdx + 1);
```

Reply:
I like it :smile: